### PR TITLE
Atomically update all linked files in a solution snapshot when one of them is updated.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes: https://github.com/dotnet/roslyn/issues/65797

Putting into feature branch as we'd prefer this to go into 17.6 to get bake time.

Adding tests now.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs [hunk 1]
```diff
@@ -1267,107 +1267,158 @@ public SolutionState WithDocumentFilePath(DocumentId documentId, string? filePat
...
-                return this;
+                var oldDocument = newState.GetRequiredDocumentState(documentId);
+                if (oldDocument.TryGetText(out var oldText) && text == oldText)
+                    continue;
+
+                newState = newState.UpdateDocumentState(oldDocument.UpdateText(text, mode), textChanged: true);
             }
 
-            return UpdateDocumentState(oldDocument.UpdateText(text, mode), textChanged: true);
+            return newState;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs:1285]: this pattern repeats so much, it would probably make sense to have a helper.  What do you think @jasonmalinowski ?

File: src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs [hunk 2]
```diff
@@ -10743,12 +10743,13 @@ public void InsertDeleteMethod_Inactive()
             // Moving inactive method declaration in a file with active statements.
 
             var srcA1 = "partial class C { void F1() { <AS:0>System.Console.WriteLine(1);</AS:0> } }";
-            var srcB1 = "partial class C { void F2() { } }";
             var srcA2 = "partial class C { void F1() { <AS:0>System.Console.WriteLine(1);</AS:0> } void F2() { } }";
+
+            var srcB1 = "partial class C { void F2() { } }";
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs:10748]: >var srcB1 = "partial class C { void F2() { } }" [](http://example.com/codeflow?start=12&length=47)

Why is this needed?

File: src/EditorFeatures/CSharpTest/EditAndContinue/Helpers/EditingTestBase.cs [hunk 3]
```diff
@@ -178,8 +178,15 @@ internal static string WrapMethodBodyWithClass(string bodySource, MethodKind kin
                  _ => "class C { void F() { " + bodySource + " } }",
              };
 
-        internal static ActiveStatementsDescription GetActiveStatements(string oldSource, string newSource, ActiveStatementFlags[] flags = null, int documentIndex = 0)
-            => new(oldSource, newSource, source => SyntaxFactory.ParseSyntaxTree(source, path: GetDocumentFilePath(documentIndex)), flags);
+        internal static ActiveStatementsDescription GetActiveStatements(
+            string oldSource, string newSource, ActiveStatementFlags[] flags = null, int documentIndex = 0)
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/Helpers/EditingTestBase.cs:182]: >oldSource [](http://example.com/codeflow?start=19&length=9)

Unnecessary change

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs [hunk 4]
```diff
@@ -1263,111 +1263,143 @@ public SolutionState WithDocumentFilePath(DocumentId documentId, string? filePat
...
-            var oldDocument = GetRequiredDocumentState(documentId);
-            if (oldDocument.TryGetText(out var oldText) && text == oldText)
-            {
-                return this;
-            }
-
-            return UpdateDocumentState(oldDocument.UpdateText(text, mode), textChanged: true);
+            return this.UpdateRelatedDocuments(
+                relatedDocuments,
+                static (newState, documentId) => newState.GetRequiredDocumentState(documentId),
```
Comment (tmat) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs:1297]: >static (newState, documentId) => newState.GetRequiredDocumentState(documentId) [](http://example.com/codeflow?start=16&length=78)

doesn't passing method group `GetRequiredDocumentState` instead of lambda work?

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs [hunk 5]
```diff
@@ -1385,50 +1417,69 @@ public SolutionState WithDocumentSyntaxRoot(DocumentId documentId, SyntaxNode ro
         /// Creates a new solution instance with the document specified updated to have the source
         /// code kind specified.
         /// </summary>
-        public SolutionState WithDocumentSourceCodeKind(DocumentId documentId, SourceCodeKind sourceCodeKind)
+        public SolutionState WithDocumentSourceCodeKind(
+            ImmutableArray<DocumentId> relatedDocuments,
+            SourceCodeKind sourceCodeKind)
```
Comment (tmat) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs:1422]: Why indent args when we have just two or three? Just makes the code consume vertical space unnecessarily.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs:1297]: no.  one is a delegate that takes two parameters, and the other is a method htat takes one.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs:1422]: my style is that once i wrap, i wrap them all.

File: src/EditorFeatures/CSharpTest/EditAndContinue/Helpers/EditingTestBase.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/EditAndContinue/Helpers/EditingTestBase.cs:182]: my general issue is when these lines go past my horizontal viewport width.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs (hunk 4 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs:1297]: Ah, C# doesn't support this while the runtime does :(

**Comment by CyrusNajmabadi** [comment]
Closing out, this didn't fit in timebox.
